shell/src/main/java/alluxio/cli/fs/command/CpCommand.java

```diff
@@ -138,36 +166,66 @@ public CopyThreadPoolExecutor(int threads, PrintStream out, AlluxioURI path) {
     }
 
     /**
-     * Prints out the message into the specified output stream in a new line, waiting if necessary
-     * for the internal message queue to have available space.
+     * Sends a message to the thread pool, if the message is of type String, then it will be
+     * displayed in stdout, if the message is of type Exception, then it will be displayed in
+     * stderr and collected into a list of exceptions, other types of messages will be ignored,
+     * this method will wait if necessary for the internal message queue to have available space.
      *
-     * @param message the message, must not be "#"
-     * @throws InterruptedException if interrupted while waiting
+     * @param message the message, either {@link String} or {@link Exception}
+     * @throws InterruptedException if interrupted while waiting to be sent
      */
-    public void println(String message) throws InterruptedException {
+    public void sendMessage(Object message) throws InterruptedException {
+      if (message instanceof Exception) {
+        mExceptions.add((Exception) message);
+      }
       mMessages.put(message);
     }
 
     /**
      * Waits until all asynchronous copy tasks succeed or fail, then shuts down the thread pool,
      * joins the printer thread, and deletes the copy destination in case of error.
      *
-     * @throws IOException when threads are interrupted or the path fails to be deleted
+     * @throws IOException summarizing all exceptions thrown in the submitted tasks and in shutdown
      */
     public void shutdown() throws IOException {
       try {
         mPool.shutdown();
-        mPool.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
-        mMessages.put(MESSAGE_DONE);
-        mPrinter.join();
+        try {
+          mPool.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+          LOG.warn("Copy thread pool is interrupted in shutdown.", e);
+          mPool.shutdownNow();
```

Comment:
reset the interrupted flag since we aren't immediately ending the thread or propagating `InterruptedException`

Reply:
what do you mean? `shutdownNow` will interrupt the internal threads.

Reply:
it interrupts the internal threads but not the current thread. If the current thread continues, it will no longer know that it was interrupted.

Basically, any time you catch `InterruptedException` and don't re-throw it, you should also call `Thread.currentThread.interrupt()`

https://programming.guide/java/handling-interrupted-exceptions.html